### ISSUE TITLE
New permissions required

### ISSUE THREAD

**Issue by binarymist** [andryou/scriptsafe#6]
Wondering why scriptsafe now requires additional permissions:
- Read and change all your data on the websites you visit
- Change your privacy-related settings

Are there release notes as to why these new permissions are required? I didn't see any specifics, may have just missed them.

Thanks for your work!

**Comment by andryou** [comment]
Thanks for the comment!

The first one: "Read and change all your data on the websites you visit" has always been required (in order for the ScriptSafe content scripts to function (e.g. blocking webbugs)).

The second one is new with the latest update, namely the new WebRTC Protection feature. https://developer.chrome.com/extensions/privacy for more info on the Chrome Privacy API.

Since all the code is available for viewing and scrutiny here, I could just tell you to look through it and find it, but I won't do that :) The exact line in ScriptSafe that relates to this is https://github.com/andryou/scriptsafe/blob/master/js/scriptsafe.js#L27 (this is the only instance where the chrome.privacy API is called)

If you have any questions or concerns, let me know!

**Comment by binarymist** [comment]
Thanks @andryou 

It's strange that chromium has only just mentioned that the first one is an additional permission now. Any thoughts/ideas as to why?

I wonder if the following line https://github.com/andryou/scriptsafe/blob/master/js/scriptsafe.js#L28 should do something that raises someones attention (throw an error for example) if webRTCIPHandlingPolicy is not available rather than swallowing the issue?

Thanks.

**Comment by andryou** [comment]
I'm not too sure; perhaps they've taken a conservative approach: if any new permission is requested by an already installed extension which already requests permissions to modify all URLs, it will re-display that permission notification.

Thanks for the note: in the next release it will check and display "This version of Chrome does not support WebRTC protection" in the Options page if the Chrome version does not support the webRTCIPHandlingPolicy feature in the chrome.privacy API.

**Comment by binarymist** [comment]
Thanks again for the work!

**Comment by derjanb** [comment]
> It's strange that chromium has only just mentioned that the first one is an additional permission now. Any thoughts/ideas as to why?

@binarymist maybe you've seen this issue: https://bugs.chromium.org/p/chromium/issues/detail?id=603822
It happens if you have ScriptSafe installed for a very long time and Chrome recently updated to version 50.

**Comment by binarymist** [comment]
Didn't see that until now @derjanb. Interesting.

**Comment by andryou** [comment]
Thanks @derjanb for the link!

**Comment by chaikhosi** [comment]
No complaints about your permissions, but it's surprising how badly Chrome handled the change. Disabling a security addon without warning, thus allowing scripts to run everywhere, until users notice an obscure prompt hidden in the settings menu is not good... it should have kept the old version and showed a more obvious prompt for users to okay the update.

Would the Chrome devs take any notice if you feed that back to them?

**Comment by LeeteqXV** [comment]
yes, second that, chaikhosi, this is incredibly bad by Chrome devs.
Should definetely keep the old version active and let us choose whether to update or not without loosing what we already have workiing and protecting us.

Any way ScriptSafe can be set to do just that, or it it entirely up to Google Chrome devs?

I am personally NOT ok with giving that privacy permission, so now I am on the lookout for an alterantive.

The irony in this:
This is just forcing us more and more over on Virtual machines such as either with Qubes-OS disposable VMs or Ubuntu with Virtualbox, OR with small containers on Ubuntu as with LXD/LXC running various parallel instances of just one browser and then skipping the security extensions alltogether on some of them...

I imagine this is not what the extension providers want.
But that is what I do if I am prevented from being in charge of my privacy options.

I wish I could install the previous version and stay there.

**Comment by LeeteqXV** [comment]
BTW: https://bugs.chromium.org/p/chromium/issues/detail?id=603822
was just fixed today, so perhaps this changes a bit for some other extensions.
That will not be the case with ScriptSafe, though, as there is actually a new requirement here.

**Comment by andryou** [comment]
@LeeteqXV if you'd like, I'd be happy to create a version of the latest ScriptSafe (with all the performance enhancements) but with the Privacy permission and related code completely removed and upload it here.

**Comment by andryou** [comment]
**I've created a branch with the legacy Permissions intact and all code related to the Privacy permission removed: https://github.com/andryou/scriptsafe/tree/legacypermissions**

**Comment by LeeteqXV** [comment]
Wow, excellent. Thanks a bunch!
FYI (for others here) I just compared the code of that "legacypermissions" version (internal version 1.0.7.4), and the only difference from 1.0.7.3 is that the webRTC stuff has been removed, which is exactly the point. Great.

**Comment by LeeteqXV** [comment]
@andryou - out of curiosity:
As a Chrome extension developer, do you have any options or flexibility in implementing the extra protection for WebRTC without requiring that extra privacy permission "by force", or are you not able to provide that unless users give away their control of those important settings? Are you in a deadlock due to architectural limitations, or is it merely a question of adding complexity (hence, time) to working around it?

(It is my impression that there is a lot to be desired when it comes to limitations in the security model of the Google Chrome extensions framework. I wonder if that is more so in Chrome, or if that is also the case from the core Chromium platform which Chrome is based on, hence same problem for other browsers like Vivaldi.com and Opera.com which also is based on Chromium?)

**Comment by andryou** [comment]
@LeeteqXV that's a very good question. As an extension developer, I'm limited by what I have access to in JavaScript and in the Chrome API. ScriptSafe relied on a workaround in blocking inline JavaScript (which did not work consistently) from inception to May 2014, when Chrome added support to block inline scripts by passing a content header policy header.

I want every feature in ScriptSafe to work as consistently and completely as possible, and WebRTC is something that can really only be controlled at the browser level, which is why I decided to utilize the Chrome API to allow the user to customize the WebRTC policy within Chrome (I'm not sure if this setting is accessible via the Settings page).

In terms of an extension ecosystem, Firefox still has a more established system and Google only recently announced they will be more strict with extensions (particularly user privacy and data). Since Chrome was released and up until now, that's ample time for malicious extension developers/newly acquired extensions to push out an update with ad/tracking/analytics malware.

In any case, I wasn't too aware of how Chrome would handle this added permission, and here we are!

I hope you find the "legacypermissions" branch useful; I will keep it synced with all the latest updates in the trunk.
